### PR TITLE
Add stdout/stderr handling to shell protocol

### PR DESCRIPTION
**Pull request by plauche**
Adds
- `stdout` message
- `stderr` message
- `exit` message
- `pid` message

Split out process handling/wrapping code into `ProcessHandler`

### CODE REVIEW

**Comment by plauche** [comment]
[Shell protocol: read from process](https://trello.com/c/WQ8TD6SH/112-shell-protocol-read-from-process)

File: libs/shell-protocol/src/messages/mod.rs [hunk 1]
```diff
@@ -32,13 +42,32 @@ pub enum Message {
...
 
 pub fn parse_message(message: ChannelMessage) -> Result<Message, ProtocolError> {
     if message.name == "spawn" {
         Ok(spawn::from_cbor(&message)?)
+    } else if message.name == "stdout" {
+        Ok(stdout::from_cbor(&message)?)
+    } else if message.name == "stderr" {
+        Ok(stderr::from_cbor(&message)?)
+    } else if message.name == "pid" {
+        Ok(pid::from_cbor(&message)?)
```
Comment (CatGarab) [inline libs/shell-protocol/src/messages/mod.rs]: Missing `exit` message

File: libs/shell-protocol/src/messages/stderr.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    };
+
+    Ok(Message::Stderr {
+        channel_id: message.channel_id,
+        data: data.to_owned(),
+    })
+}
+
+/// Stderr -> CBOR
+pub fn to_cbor(channel_id: u32, data: Option<&str>) -> Result<Vec<u8>, ProtocolError> {
```
Comment (CatGarab) [inline libs/shell-protocol/src/messages/stderr.rs]: What's the purpose or implication of having an empty stderr/stdout message? (data = None)

File: libs/shell-protocol/src/messages/stderr.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        })?,
+    )
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use channel_protocol::parse_message;
+    use serde_cbor::de;
+
```
Comment (CatGarab) [inline libs/shell-protocol/src/messages/stderr.rs]: nit: Add a test case for empty data

File: libs/shell-protocol/src/process.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                            err,
+                        })
+                    }
+                }
+            }
+            None => Ok(None),
+        }
+    }
+
+    /// Retrieve id of process
```
Comment (CatGarab) [inline libs/shell-protocol/src/process.rs]: nit: capitalize ID

File: libs/shell-protocol/src/protocol.rs [hunk 5]
```diff
@@ -43,27 +48,58 @@ impl Protocol {
...
+    pub fn message_engine<F>(&mut self, pump: F, timeout: Duration) -> Result<(), ProtocolError>
     where
         F: Fn(Duration) -> Result<ChannelMessage, ProtocolError>,
     {
         loop {
+            // Check child for new stdout data
+            // Reading from stdout/stderr is currently a blocking operation...
+            // This code should probably get refactored into something akin
+            // to an event loop with non-blocking io.
+            if let Some(process) = self.process.as_mut().as_mut() {
```
Comment (CatGarab) [inline libs/shell-protocol/src/protocol.rs]: Why do you need two `as_mut()` calls?

File: libs/shell-protocol/src/process.rs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    ///
+    /// This will block as long as the process is running.
+    /// A return value of `None` indicates the stream is
+    /// no longer available and likewise the process
+    /// is likely no longer alive.
+    pub fn read_stdout(&mut self) -> Result<Option<String>, ProtocolError> {
+        match self.stdout_reader {
+            Some(ref mut stdout_reader) => {
+                let mut data = String::new();
+                match stdout_reader.read_line(&mut data) {
```
Comment (CatGarab) [inline libs/shell-protocol/src/process.rs]: Will this block indefinitely until a line's worth of data is available? (and the same for stderr)
If so, could you use something like [this crate](https://docs.rs/timeout-readwrite/0.1.0/timeout_readwrite/) to add a timeout?

File: libs/shell-protocol/src/messages/stderr.rs (hunk 2 above)
Comment (plauche) [reply libs/shell-protocol/src/messages/stderr.rs]: The empty message indicates that the process's stderr/stdout pipe has been closed.

File: libs/shell-protocol/test.sh [hunk 7]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+echo hello
+echo world
```
Comment (CatGarab) [inline libs/shell-protocol/test.sh]: nit: test file which probably doesn't need to be included
